### ISSUE TITLE
Make reproducible builds the default

### ISSUE THREAD

**Issue by gredler** [gradle/gradle#28806]
### Expected Behavior

One important foundation of the open source ecosystem is trust. Trust that the code does what it says on the tin, and trust that the project developers are managing the project in good faith. One area where this trust is particularly important is in the gap between source code and binaries, where users have to trust that the code they see tagged as version X matches the binary distributed as version X.

In the spirit of "trust but verify", the [Reproducible Builds](https://reproducible-builds.org/) project tries to close this gap across a number of technology stacks, including Java, by [verifying](https://github.com/jvm-repo-rebuild/reproducible-central) the reproducibility of Java libraries on Maven Central.

The Gradle team recognized the importance of this traceability in 2019 and migrated Gradle itself to reproducible builds (see #8096, #8541). Unfortunately, most developers currently have to opt-in to reproducible builds by customizing a few [documented configuration settings](https://docs.gradle.org/current/userguide/working_with_files.html#sec:reproducible_archives), and possibly customizing a few undocumented configuration settings as well (see #10900). It would be ideal if the default settings for these few configuration items provided strong security guarantees, allowing users to opt-out of better security rather than requiring them to opt-in.

The expected behavior would be that all `AbstractArchiveTask`s use the following configuration by default:

```
preserveFileTimestamps = false
reproducibleFileOrder = true
dirMode = 0755
fileMode = 0644
```

### Current Behavior (optional)

The current behavior is to use the build machine's current date/time, file iteration order, and file permission settings. This is configuration that will vary across build machines and even across build attempts on a single machine.

### Context

The recent [xz backdoor](https://securelist.com/xz-backdoor-story-part-1/112354/) highlights the dangers of this source/binary gap and possible outcomes when user trust is undermined in this area. As both a maintainer of open source projects and a consumer of open source projects, my systems would be safer if Gradle changed these defaults, migrating a large chunk of the Java ecosystem to reproducible builds by default.

**Comment by ov7a** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by sdavids** [comment]
```kotlin
tasks.withType<AbstractArchiveTask>().configureEach {
  isPreserveFileTimestamps = false
  isReproducibleFileOrder = true
  filePermissions {
    unix(644)
  }
  dirPermissions {
    unix(755)
  }
}
```

**Comment by jjohannes** [comment]
**In case a source set contains a `module-info.java`**

Then, the corresponding `JavaCompile` task (or the `javaCompiler` property of the task) should track _major.minor.patch_ version of the used Java toolchain. The version is compiled into the `module-info.class` files as part of the  _requires_ that point at JDK Java Modules (like `java.base`). Compiling the same `module-info.java` with different  _major.minor.patch_ versions of a JDK thus creates different results.

Example of the same `module-info.java` compiled with `17.0.9` vs `17.0.3`:

<img width="640" alt="mi" src="https://github.com/gradle/gradle/assets/1963746/fa0bd5cb-4aac-451c-b789-65da93905010">

Related: https://github.com/gradle/gradle/issues/16628#issuecomment-2056923926

**Comment by gredler** [comment]
> Even if a source set does not contain a module-info.java, in order to get reproducible builds it is important to be able to fix the toolchain used

True, at one point the internal JDK configuration of the font engine changed slightly between minor version releases, and the https://github.com/woo-j/OkapiBarcode unit tests would pass when running against OpenJDK 11.0.n but fail when running against 11.0.n+1 because the tests verify barcode rendering pixel by pixel (including the text under the barcodes). I guess that's another example of a build that would benefit from a more granular toolchain versioning mechanism.

**Comment by hfhbd** [comment]
Is there any Gradle plugin to create the `. buildinfo` and publish it too?

**Comment by gredler** [comment]
> Is there any Gradle plugin to create the `.buildinfo` and publish it too?

I'm not aware of one, but IMO the tooling could easily (at least in many cases) make these files obsolete. They generally specify the Maven coordinates (which you would already have if you are iterating through Maven Central or if you are told to check a specific artifact), the git repository (which is usually provided in the POM metadata anyway), the git tag naming scheme (this one is usually not available anywhere else, but there are a handful of tag name conventions that most projects follow and each could be checked for a match), the build tool used (easily guessed based on the repository contents), the JDK used (probably already available in the toolchain metadata provided in the build file), the newline character used in files (this one is a mystery to me), a custom file umask (AFAIK it just provides an extra build resilience check), and the actual build command to run (it's pretty much the same set of Gradle flags and tasks for most projects, the only real variance that I've seen is project-specific config to disable signing when publishing to the local Maven repository).

If Gradle could provide a standard flag for disabling artifact signing, or change the plugin defaults so that publishing to the local Maven repository does not trigger signing, it seems to me that the .buildinfo files used by the Reproducible Builds project to verify reproducibility would only be necessary for the most complex of builds.

**Comment by hboutemy** [comment]
no, please don't publish `.buildinfo`: it's an idea from 5 years ago that finally does not help

**Comment by lptr** [comment]
While making reproducible builds the default is not something we are planning to do in the foreseeable future, tracking the Java version that makes it into the `module-info.class` sounds like something we should do. @jjohannes could you create an issue for that please?

**Comment by jjohannes** [comment]
@lptr there is an issue for that from my point of view (and a suggestion for a solution): https://github.com/gradle/gradle/issues/16628#issuecomment-2097540617 

The JavaToolchainSpec is already input of all relevant tasks. It "just" needs to support specifying the complete Java version.

**Comment by gredler** [comment]
@lptr Thank you for the update. There are four archive task attributes that currently need to be customized for reproducible builds, have you discussed the possibility of changing one or two of these defaults to get the ecosystem closer to reproducibility? (i.e. not letting "perfect" be the enemy of "better")

**Comment by lptr** [comment]
I think we sould come back here and reevaluate our position on reproducible archives. This is not an area we are currently focusing on though, so I can't promise exactly when.

**Comment by Omico** [comment]
> ```kotlin
> tasks.withType<AbstractArchiveTask>().configureEach {
>   isPreserveFileTimestamps = false
>   isReproducibleFileOrder = true
>   filePermissions {
>     unix(644)
>   }
>   dirPermissions {
>     unix(755)
>   }
> }
> ```

This doesn't work as expected when using Kotlin because Kotlin doesn't support octal literals. See also https://youtrack.jetbrains.com/issue/KT-70717.

Suggest using symbolic notation instead for better readability!

```
tasks.withType<AbstractArchiveTask>().configureEach {
    isPreserveFileTimestamps = false
    isReproducibleFileOrder = true
    dirPermissions {
        // 755
        unix("rwxr-xr-x")
    }
    filePermissions {
        // 644
        unix("rw-r--r--")
    }
}
```

**Comment by sdavids** [comment]
https://github.com/gradle/gradle/blob/035834c2fa7b88422635cde176fe9c7a6f6b8bc7/platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/DefaultConfigurableFilePermissions.java#L96-L100

https://github.com/gradle/gradle/blob/035834c2fa7b88422635cde176fe9c7a6f6b8bc7/platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFilePermissions.java#L32-L35

https://github.com/gradle/gradle/blob/035834c2fa7b88422635cde176fe9c7a6f6b8bc7/platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/DefaultConfigurableUserClassFilePermissions.java#L38-L42

https://github.com/gradle/gradle/blob/035834c2fa7b88422635cde176fe9c7a6f6b8bc7/platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/AbstractUserClassFilePermissions.java#L33-L35


```shell
jshell --feedback=silent - <<< 'System.out.println((((755 & 0_700) >> 6) & 4) >> 2 == 1)'
false
 ~ % jshell --feedback=silent - <<< 'System.out.println((((0755 & 0_700) >> 6) & 4) >> 2 == 1)'
true
```

I think the Javadoc of [ConfigurableFilePermissions.html#unix(int)](https://docs.gradle.org/current/javadoc/org/gradle/api/file/ConfigurableFilePermissions.html#unix(int)) could be better; it currently implies that decimal values are OK.

**Comment by sdavids** [comment]
#30369

**Comment by jjohannes** [comment]
We published a plugin under the GradleX umbrella that does some of the things discussed here:
- `AbstractArchiveTask` configurations 
- Setting encoding to `UTF-8` for Compile and JavaDoc tasks

All details: https://gradlex.org/reproducible-builds/

(Unfortunately, the [define the exact Java Version in toolchains](https://github.com/gradle/gradle/issues/16628#issuecomment-2097540617) topic cannot be solved in a plugin.)

**Comment by lptr** [comment]
Shall we rename this to "Make reproducible archives the default"? If so, I think we should handle #30826 and #30827 under this issue.

**Comment by lptr** [comment]
https://github.com/gradle/gradle/issues/14819 is another thing we could put under this umbrella.

**Comment by asodja** [comment]
A late update to the issue. By fixing https://github.com/gradle/gradle/issues/30871 with Gradle 9.0.0 we are closer to the reproducible builds. 

The behaviour from the issue description is implemented:
> The expected behavior would be that all AbstractArchiveTasks use the following configuration by default:
> preserveFileTimestamps = false
> reproducibleFileOrder = true
> dirMode = 0755
> fileMode = 0644

But there are still other details that needs to be addressed to make builds fully reproducible, like reproducible toolchains etc. So we can keep this issue open.
